File: framework/languages/java/e2e/RegistrationCodec/Client/src/main/java/systems/zlink/e2e/registrationcodec/client/Scenarios/RcA5FilterOrderingScenario.java
package systems.zlink.e2e.registrationcodec.client.Scenarios;

import systems.zlink.e2e.registrationcodec.client.Support.ScenarioAssert;
import systems.zlink.e2e.registrationcodec.client.Support.ScenarioContext;
import systems.zlink.e2e.registrationcodec.shared.Contracts;

public final class RcA5FilterOrderingScenario {
    private RcA5FilterOrderingScenario() {
    }

    public static void run(ScenarioContext context) {
        Contracts.EchoRes filtered = context.server().post("/registration/filter-order").submit(Contracts.EchoRes.class).toCompletableFuture().join().body();
        ScenarioAssert.ensure("echo:filter-order-request".equals(filtered.value()), "RC-A5 request mismatch");
        ScenarioAssert.waitForFilterOrder(context.evidence(), "filter-order-request");
        System.out.println("scenario RC-A5 passed");
    }
}

File: framework/languages/java/e2e-kotlin/RegistrationCodec/Client/src/main/kotlin/systems/zlink/e2e/kotlin/registrationcodec/client/Scenarios/RcA5FilterOrderingScenario.kt
package systems.zlink.e2e.kotlin.registrationcodec.client.scenarios

import systems.zlink.e2e.kotlin.registrationcodec.EchoManualRes
import systems.zlink.e2e.kotlin.registrationcodec.client.support.ScenarioAssert
import systems.zlink.e2e.kotlin.registrationcodec.client.support.ScenarioHttpClient

class RcA5FilterOrderingScenario(
    private val server: ScenarioHttpClient,
    private val assert: ScenarioAssert,
) {
    fun run() {
        val filtered = server.post<EchoManualRes>("/registration/filter-order")
        assert.that(filtered.value == "echo:filter-order-request", "RC-A5 request mismatch")
        val expected = listOf(
            "first-before",
            "second-before",
            "second-after",
            "first-after",
        )
        val actual = server.waitForEvidence("Filter", "EchoManualReq", "first-after")
            .filter { it.marker == "Filter" && it.packetName == "EchoManualReq" }
            .map { it.value }
        assert.that(
            actual.size >= expected.size && actual.takeLast(expected.size) == expected,
            "RC-A5 filter order mismatch: $actual",
        )
        println("scenario RC-A5 passed")
    }
}
